src/Widgets/calendar/GridDay.vala

```diff
@@ -70,20 +69,70 @@ public class DateTime.Widgets.GridDay : Gtk.EventBox {
 
         label.name = "date";
 
-        add (label);
+        main_grid = new Gtk.Grid ();
+        main_grid.hexpand = true;
+        main_grid.halign = Gtk.Align.CENTER;
+        main_grid.valign = Gtk.Align.CENTER;
+        main_grid.orientation = Gtk.Orientation.VERTICAL;
+        main_grid.attach (label, 0, 0);
+
+        add (main_grid);
+        halign = Gtk.Align.CENTER;
         show_all ();
 
         // Signals and handlers
         button_press_event.connect (on_button_press);
         key_press_event.connect (on_key_press);
         scroll_event.connect ((event) => {return Util.on_scroll_event (event);});
+
+        Widgets.CalendarModel.get_default ().events_added.connect (update_event_days);
+        Widgets.CalendarModel.get_default ().events_updated.connect (update_event_days);
+        Widgets.CalendarModel.get_default ().events_removed.connect (update_event_days);
     }
 
     public void update_date (GLib.DateTime date) {
         this.date = date;
         label.label = date.get_day_of_month ().to_string ();
     }
 
+    public async void update_event_days () {
+        GLib.Idle.add (() => {
+            var events = Widgets.CalendarModel.get_default ().get_events (date);
+            var event_dot_grid = new Gtk.Grid ();
+            event_dot_grid.column_homogeneous = true;
+            if (event_dot_grid != null) {
+                event_dot_grid.destroy ();
+            }
+            if (events.size != 0) {
+                if (events.size <= 4) {
+                    foreach (var e in events) {
+                        if (e != null) {
+                            var event_dot = new Gtk.Image.from_icon_name ("pager-checked-symbolic", Gtk.IconSize.MENU);
+                            event_dot.visible = true;
+                            event_dot.pixel_size = 6;
+                            event_dot.halign = Gtk.Align.CENTER;
+
+                            Util.set_event_calendar_color (e.cal, event_dot);
+
+                            event_dot_grid.add (event_dot);
+                        }
+                    }
+                } else if (events.size > 4) {
+                    var event_bar = new Gtk.Image.from_icon_name ("list-remove-symbolic", Gtk.IconSize.MENU);
+                    event_bar.visible = true;
+                    event_bar.pixel_size = 6;
```

Comment:
If you're setting all the same properties, wouldn't it be better to create the image and set its properties first, then set a gicon in the conditional?